---
title: 'Reanalysis of 201113 Mouse CRPS Combined: DADA2 Pathogen Server with SILVA'
author: "Rachel Rodgers"
date: '`r format(Sys.Date(), "%B, %d, %Y")`'
output: html_document
---

```{r load-libraries, message=FALSE}
library("dada2")
library("msa")
library("phangorn")
library("phyloseq")
library("tidyverse")
```

```{r set-directories}
#----- Raw Data & Filtered Data Paths -----#
inputDirCohort1 <- "../data/raw_data/190809_Lara_16S"
inputDirCohort2 <- "../data/raw_data/201910_LaraC_V4_Mouse_Human_CRPS"
filteredDir <- "../data/filtered"

#----- Taxonomy Database Paths -----#
taxonomyDBPath <- "/mnt/pathogen1/rrodgers/databases/dada2_taxonomy"

# Silva
silvaDB <- file.path(taxonomyDBPath, "silva_v138.1/silva_nr99_v138.1_train_set.fa.gz")
silvaDBSpecies <- file.path(taxonomyDBPath, "silva_v138.1/silva_species_assignment_v138.1.fa.gz")
```

This data set contains two cohorts:

  * 190809_Lara_16S (112 samples)
  * 201019_LaraC_V4_Mouse_Human_CRPS (100 samples)
  
These combined cohorts were previously analyzed using the RDP taxa assignments
in the project 201113_Mouse_CRPS_Combined.

In this script, I will be annotating the taxa in these cohorts with SILVA.

The raw data directories contain samples not needed in the current study.
Here I will use the data contained in the compiled metadata to select just the
required samples.

```{r select-samples}
sampleData <- read.delim("../data/metadata/201113_Mouse_CRPS_combinedMetadata.txt",
                         stringsAsFactors = FALSE, check.names = FALSE)

#----- Cohort 1 -----#
# cohort 1 samples: data/raw_data/190809_Lara_16S

# Build pattern for matching cohort 1 R1 files:
cohort1Patterns <- sampleData %>% 
  filter(cohort == "cohort_1") %>% 
  pull(sample)
cohort1PatternsVec <- paste0("^Baldridge_", cohort1Patterns, "_.?")
cohort1PatternsStr <- paste(cohort1PatternsVec, collapse = "|")

allCohort1R1Files <- sort(list.files(inputDirCohort1,
                                     pattern = "_R1_001.fastq.gz")) # 191 total

useCohort1FilesLogical <- map_lgl(.x = allCohort1R1Files,
                                  .f = ~ grepl(pattern = cohort1PatternsStr,
                                               x = .x))
useCohort1Files <- allCohort1R1Files[useCohort1FilesLogical] # 112

#----- Cohort 2 -----#
# cohort 2 samples: data/raw_data/201910_LaraC_V4_Mouse_Human_CRPS

# Build pattern for matching cohort 2 R1 files:
cohort2Patterns <- sampleData %>% 
  filter(cohort == "cohort_2") %>% 
  pull(sample)

cohort2PatternsVec <- paste0("^", cohort2Patterns, "_.?")
cohort2PatternsStr <- paste(cohort2PatternsVec, collapse = "|")

allCohort2R1Files <- sort(list.files(inputDirCohort2, 
                                     pattern = "_R1_001.fastq.gz"))# 123 total

useCohort2FilesLogical <- map_lgl(.x = allCohort2R1Files,
                                  .f = ~ grepl(pattern = cohort2PatternsStr,
                                               x = .x))
useCohort2Files <- allCohort2R1Files[useCohort2FilesLogical] # 100

#----- Combine to make list of files names with path -----#
allInputFiles <- c(useCohort1Files, useCohort2Files)

allInputFilePaths <- c(file.path(inputDirCohort1, useCohort1Files),
                       file.path(inputDirCohort2, useCohort2Files))
```

Check the raw read quality of the combined mouse cohorts.

```{r rawQualPlot, fig.width=9, fig.height=6}
rawQualPlot <- plotQualityProfile(allInputFilePaths, aggregate = TRUE)
rawQualPlot
```

Combined data from across two runs shows decreased quality around 225 bp.

```{r filter-and-trim, eval=FALSE}
filterAndTrim(fwd = allInputFilePaths,
              filt = file.path(filteredDir, allInputFiles),
              trimLeft = 10, maxEE = 2, truncQ = 11,
              maxN = 0, rm.phix = TRUE, compress = TRUE, verbose = TRUE,
              truncLen = 225)
```

Infer error rates.

```{r filteredFiles}
# Get list of filtered files and assign sample names to them
filteredFiles <- list.files(filteredDir, pattern = "fastq.gz", full.names = TRUE)

# Get the sample names by removing the .fastq and path 
sampleNames <- map_chr(basename(filteredFiles), 
                       ~ str_remove(string = .x,
                                    pattern = "_[:alpha:]{12}_[:alpha:]{12}.+_L001_R1_001.fastq.gz"))
# Assign the sample names to the filteredFiles vector
names(filteredFiles) <- sampleNames
```

```{r errF, fig.width=9, fig.height=6}
set.seed(12729533)

errF <- learnErrors(filteredFiles, multithread = TRUE, nbases = 1e9)

errorPlot <- plotErrors(errF, nominalQ = TRUE)
errorPlot
```

Dereplicate.

```{r dereplication}
# Create a list that will hold dereplication objects for each sample
singles <- vector("list", length(sampleNames))
names(singles) <- sampleNames

# Populate the list
for(sample in sampleNames) {
  derepF <- derepFastq(filteredFiles[[sample]])
  singles[[sample]] <- dada(derepF, err = errF, multithread = TRUE)
}
```

```{r sequence-table}
# Construct sequence table and remove chimeras
sequenceTable <- makeSequenceTable(singles)
sequenceTableNoChimeras <- removeBimeraDenovo(sequenceTable,
                                              multithread = TRUE)
```

```{r assign-taxonomy}
taxaRankNamesFull <- c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species")
taxaRankNamesTrunc <- c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus")

# Silva
taxaSilva <- assignTaxonomy(sequenceTableNoChimeras, silvaDB, multithread = TRUE)
#colnames(taxaSilva) <- taxaRankNamesTrunc
taxaSilvaPlus <- addSpecies(taxaSilva, silvaDBSpecies)
```

```{r construct-phylogenetic-tree}
# Get the sequences from the sequence table
seqs <- getSequences(sequenceTableNoChimeras)
names(seqs) <- seqs
# Multiple sequence alignment
mult <- msa(seqs, method = "ClustalW", type = "dna", order = "input")
# Convert MSA to phyDAT format
phangAlign <- as.phyDat(mult, type = "dna", order = "input")
# Compute pairwise distances on phangAlign
dm <- dist.ml(phangAlign)
treeNJ <- NJ(dm)
# Compute likelihood of tree
fit <- pml(tree = treeNJ, data = phangAlign)
fitGTR <- update(fit, k = 4, inv = 0.2)
fitGTR <- optim.pml(fitGTR, model = "GTR", optInv = TRUE, optGamma = TRUE,
                    rearrangement = "stochastic", 
                    control = pml.control(trace = 0))

save.image()
```

```{r build-phyloseq-objects}
# silva
ps0.silva <- phyloseq(otu_table(sequenceTableNoChimeras, taxa_are_rows = FALSE),
                      tax_table(taxaSilvaPlus), phy_tree(fitGTR$tree))
ps0.silva # 936 x 212
saveRDS(ps0.silva, "../data/RDataObjects/physeqObjects/ps0.silvaNR99v138.1_single.RDS")
```

```{r save-data}
save.image("./silva/Reanalysis_201113MouseCRPSCombined_dada2.RData") # read back w/ load() or attach()
writeLines(capture.output(sessionInfo()), 
           "./silva/Reanalysis_201113MouseCRPSCombined_dada2_session_info.txt")
Sys.Date()
getwd()
sessionInfo()
```
